File: Problem144/Solution.java
import java.util.*;

public class Solution {
	public List<Integer> preorderTraversal(TreeNode root) {
		List<Integer> list = new ArrayList<Integer>();
		populateInOrder(root, list);

		return list;
	}

	private void populateInOrder(TreeNode root, List<Integer> list) {
		if (root != null) {
			list.add(root.val);
			populateInOrder(root.left, list);
			populateInOrder(root.right, list);
			return;
		}
	}

	public static class TreeNode {
		int val;
		TreeNode left;
		TreeNode right;

		TreeNode(int x) {
			val = x;
		}
	}
}

File: Problem144/Solution.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func preorderTraversal(root *TreeNode) []int {
	var list []int
	populatePreOrder(root, &list)
	return list
}

func populatePreOrder(root *TreeNode, list *[]int) {
	if root != nil {
		*list = append(*list, root.Val)
		populatePreOrder(root.Left, list)
		populatePreOrder(root.Right, list)
	}
}
